crates/model/src/order.rs

```diff
@@ -388,28 +411,51 @@ impl OrderCreation {
     }
 }
 
-impl Default for OrderCreation {
-    // Custom implementation to make sure the default order creation is valid.
+// Note that the order of the variants is important for deserialization.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[serde(untagged)]
+pub enum OrderCreationAppData {
+    /// Hash is inferred from full app data and validated against expectation.
+    Both {
+        #[serde(rename = "appData")]
+        full: String,
+        #[serde(rename = "appDataHash")]
+        expected: AppDataHash,
+    },
+    /// Backward compatible app data hash.
+    Hash {
+        #[serde(rename = "appData")]
+        hash: AppDataHash,
+    },
+    /// Hash is inferred from full app data.
+    Full {
+        #[serde(rename = "appData")]
+        full: String,
+    },
+}
+
+impl Default for OrderCreationAppData {
     fn default() -> Self {
-        Self {
-            data: OrderData {
-                valid_to: u32::MAX,
-                ..Default::default()
-            },
-            from: None,
-            signature: Signature::Eip712(EcdsaSignature::non_zero()),
-            quote_id: None,
+        Self::Hash {
+            hash: Default::default(),
         }
     }
 }
 
-impl From<Order> for OrderCreation {
-    fn from(order: Order) -> Self {
-        OrderCreation {
-            data: order.data,
-            from: Some(order.metadata.owner),
-            signature: order.signature,
-            quote_id: None,
+impl From<AppDataHash> for OrderCreationAppData {
+    fn from(hash: AppDataHash) -> Self {
+        Self::Hash { hash }
+    }
+}
+
+impl OrderCreationAppData {
+    /// Returns the app data hash. Does not validate expectation.
+    pub fn hash(&self) -> AppDataHash {
+        match self {
+            Self::Hash { hash } => *hash,
+            Self::Full { full } | Self::Both { full, .. } => {
+                AppDataHash(app_data_hash::hash_full_app_data(full.as_bytes()))
```

Comment:
Should this error for `Self::Both` if `app_data_hash::hash_full_app_data(full) != hash`?

I.e., is there a reason why the validation logic can't move here?

Reply:
The expectation is a quirk of our order creation validation. If a user provides an expectation, we may validate it. I find it weird to handle that logic here.

Reply:
But then the `OrderCreationAppData -> AppDataHash` has some ambiguous semantics which feels a bit weird.

Reply:
In my mind, the abstraction is an `OrderCreation` that we get from a user, and that data can sometimes not correctly convert to an `OrderData` (the validation). In that sense, having `OrderCreation::order_data` and `OrderCreationAppData::hash` potentially return errors reflects that well.

Reply:
The semantics of this method are to return the smart contract order's app data field for this order.

If you want to move this validation here then how do you justify not moving all of the order validation code into here? We could also validate that valid_to is not in the past and so on. But that doesn't make sense to me because it is one more level detached from the structural representation of the api types that the model crate is supposed to provide.

Reply:
I think that is a different kind of validation though - validating the app data is checking that the `OrderCreation` data is consistent (like checking signatures for example, which is in this module AFAIR), while the `validTo` check is validating "order rules" - an expired order is a valid `OrderData`, it just can't be used and is thus rejected by the API